File: src/me/shkschneider/skeleton/helpers/ScreenHelper.java
package me.shkschneider.skeleton.helpers;

import android.app.Activity;
import android.graphics.Rect;
import android.os.PowerManager;
import android.util.TypedValue;
import android.view.Window;
import android.view.WindowManager;

import me.shkschneider.skeleton.SkeletonApplication;

@SuppressWarnings("unused")
public final class ScreenHelper {

    public static final int DENSITY_LDPI = 120;
    public static final int DENSITY_MDPI = 160;
    public static final int DENSITY_HDPI = 240;
    public static final int DENSITY_XHDPI = 320;
    public static final int DENSITY_XXHDPI = 480;
    public static final int DENSITY_XXXHDPI = 640;
    public static final int DENSITY_TV = 213;

    public static boolean wakeLock(final Activity activity) {
        if (activity == null) {
            LogHelper.warning("Activity was NULL");
            return false;
        }

        activity.getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
        return true;
    }

    public static boolean on() {
        return ((PowerManager) SystemHelper.systemService(SystemHelper.SYSTEM_SERVICE_POWER_SERVICE)).isScreenOn();
    }

    public static int density() {
        return SkeletonApplication.CONTEXT.getResources().getDisplayMetrics().densityDpi;
    }

    public static int height() {
        return SkeletonApplication.CONTEXT.getResources().getDisplayMetrics().heightPixels;
    }

    public static int statusBarHeight(final Window window) {
        if (window == null) {
            LogHelper.warning("Window was NULL");
            return 0;
        }

        final Rect rect = new Rect();
        window.getDecorView().getWindowVisibleDisplayFrame(rect);
        return rect.top;
    }

    public static int actionBarHeight(final Window window) {
        if (window == null) {
            LogHelper.warning("Window was NULL");
            return 0;
        }

        int top = window.findViewById(Window.ID_ANDROID_CONTENT).getTop();
        return top - statusBarHeight(window);
    }

    public static int width() {
        return SkeletonApplication.CONTEXT.getResources().getDisplayMetrics().widthPixels;
    }

    public static int orientation() {
        final WindowManager windowManager = (WindowManager) SystemHelper.systemService(SystemHelper.SYSTEM_SERVICE_WINDOW_SERVICE);
        return windowManager.getDefaultDisplay().getRotation();
    }

    public static int pixelsFromDp(final float dp) {
        if (dp <= 0F) {
            LogHelper.warning("Dp was too low");
            return 0;
        }

        return (int) (TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp, SkeletonApplication.CONTEXT.getResources().getDisplayMetrics()));
    }

}

File: android/src/main/kotlin/me/shkschneider/skeleton/android/helper/ScreenHelper.kt
package me.shkschneider.skeleton.android.helper

import android.content.Context
import android.graphics.Point
import android.util.DisplayMetrics
import android.view.Surface
import android.view.Window
import android.view.WindowManager
import androidx.annotation.FloatRange
import me.shkschneider.skeleton.android.R
import me.shkschneider.skeleton.android.app.ApplicationHelper
import me.shkschneider.skeleton.android.util.Metrics
import me.shkschneider.skeleton.android.provider.SystemServices
import me.shkschneider.skeleton.kotlin.log.Logger
import kotlin.math.pow
import kotlin.math.sqrt

object ScreenHelper {

    const val DENSITY_LDPI = 120
    const val DENSITY_MDPI = 160
    const val DENSITY_TVDPI = 213 // MDPI * 1.33
    const val DENSITY_HDPI = 240
    const val DENSITY_XHDPI = 320
    const val DENSITY_XXHDPI = 480
    const val DENSITY_XXXHDPI = 640

    const val BRIGHTNESS_RESET = (-1.1).toFloat()
    const val BRIGHTNESS_MIN = 0.1.toFloat()
    const val BRIGHTNESS_MAX = 1.1.toFloat()

    const val ROTATION_0 = Surface.ROTATION_0
    const val ROTATION_90 = Surface.ROTATION_90
    const val ROTATION_180 = Surface.ROTATION_180
    const val ROTATION_270 = Surface.ROTATION_270

    // Does NOT @RequiresPermission(Manifest.permission.WAKE_LOCK)
    fun wakeLock(window: Window) {
        window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
    }

    // Does NOT @RequiresPermission(Manifest.permission.WAKE_LOCK)
    fun wakeUnlock(window: Window) {
        window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
    }

    fun on(): Boolean? {
        val powerManager = SystemServices.powerManager()
        return powerManager?.isInteractive
    }

    fun brightness(window: Window): Float {
        return window.attributes.screenBrightness
    }

    fun brightness(window: Window, @FloatRange(from = BRIGHTNESS_RESET.toDouble(), to = BRIGHTNESS_MAX.toDouble()) brightness: Float) {
        val layoutParams = window.attributes
        layoutParams.screenBrightness = if (brightness < BRIGHTNESS_MIN) BRIGHTNESS_RESET else brightness
        window.attributes = layoutParams
    }

    // <https://github.com/eyeem/deviceinfo>
    fun inches(): Float {
        val windowManager = SystemServices.windowManager()
        val displayMetrics = DisplayMetrics()
        val display = windowManager?.defaultDisplay
        display?.getMetrics(displayMetrics)
        val point = Point(1, 1)
        display?.getRealSize(point)
        val w = point.x / displayMetrics.xdpi
        val h = point.y / displayMetrics.ydpi
        val inches = sqrt(w.toDouble().pow(2.0) + h.toDouble().pow(2.0)).toFloat()
        Logger.verbose("Inches: $inches\"")
        return inches
    }

    fun height(): Int {
        return Metrics.displayMetrics().heightPixels
    }

    fun width(): Int {
        return Metrics.displayMetrics().widthPixels
    }

    fun statusBarHeight(): Int {
        val resId = ApplicationHelper.resources.getIdentifier("status_bar_height", "dimen", "android")
        if (resId > 0) {
            return ApplicationHelper.resources.getDimensionPixelSize(resId)
        }
        return ApplicationHelper.resources.getDimensionPixelSize(R.dimen.sk_statusBar)
    }

    fun navigationBarHeight(): Int {
        val resId = ApplicationHelper.resources.getIdentifier("navigation_bar_height", "dimen", "android")
        if (resId > 0) {
            return ApplicationHelper.resources.getDimensionPixelSize(resId)
        }
        return ApplicationHelper.resources.getDimensionPixelSize(R.dimen.sk_statusBar)
    }

    fun orientation(context: Context): Int {
        return context.resources.configuration.orientation
    }

    fun rotation(): Int? {
        return SystemServices.windowManager()?.defaultDisplay?.rotation
    }

}